bookwyrm/views/follow.py

```diff
@@ -84,3 +94,68 @@ def delete_follow_request(request):
 
     follow_request.delete()
     return redirect(f"/user/{request.user.localname}")
+
+
+def ostatus_follow_request(request):
+    """prepare an outgoing remote follow request"""
+    uri = urllib.parse.unquote(request.GET.get("acct"))
+    username_parts = re.search(
+        r"(?:^http(?:s?):\/\/)([\w\-\.]*)(?:.)*(?:(?:\/)([\w]*))", uri
+    )
+    account = f"{username_parts[2]}@{username_parts[1]}"
+    user = handle_remote_webfinger(account)
+    error = None
+
+    if user is None or user == "":
+        error = "ostatus_subscribe"
+
+    # don't do these checks for AnonymousUser before they sign in
+    if request.user.id:
+
+        # you have blocked them so you probably don't want to follow
+        if hasattr(request.user, "blocks") and user in request.user.blocks.all():
+            error = "is_blocked"
+        # they have blocked you
+        if hasattr(user, "blocks") and request.user in user.blocks.all():
+            error = "has_blocked"
+        # you're already following them
+        if hasattr(user, "followers") and request.user in user.followers.all():
+            error = "already_following"
+        # you're not following yet but you already asked
+        if (
+            hasattr(user, "follower_requests")
+            and request.user in user.follower_requests.all()
+        ):
+            error = "already_requested"
+
+    data = {"account": account, "user": user, "error": error}
+
+    return TemplateResponse(request, "ostatus/subscribe.html", data)
+
+
+@login_required
+def ostatus_follow_success(request):
+    """display success message for remote follow"""
+    user = get_user_from_username(request.user, request.GET.get("following"))
+    data = {"account": user.name, "user": user, "error": None}
+    return TemplateResponse(request, "ostatus/success.html", data)
+
+
+def remote_follow_page(request):
+    """Display remote follow page"""
+    user = get_user_from_username(request.user, request.GET.get("user"))
+    data = {"user": user}
+    return TemplateResponse(request, "ostatus/remote_follow.html", data)
+
+
+@require_POST
+def remote_follow(request):
+    """direct user to follow from remote account using ostatus subscribe protocol"""
+    remote_user = request.POST.get("remote_user")
+    template = subscribe_remote_webfinger(remote_user)
+    if template is None:
+        data = {"account": remote_user, "user": None, "error": "remote_subscribe"}
```

Comment:
Would it be possible to create separate errors for when a user isn't found on the server vs when the server isn't equipped for remote follow?

Reply:
Oh yeah ghat would be much clearer for the BookWyrm user. May need to add some additional code to differentiate no server / user not found / user found but service doesn't use ostatus subscribe, though we'd use friendlier descriptions